testctrl/proto/test_sessions.proto

```diff
@@ -0,0 +1,185 @@
+    UNSPECIFIED = 0;
+
+    // Orchestrates the test amongst workers and reports results.
+    DRIVER = 1;
+
+    // Processes incoming requests from client containers.
+    SERVER = 2;
+
+    // Sends outgoing requests to a server container.
+    CLIENT = 3;
+  }
+
+  // The container image and a version as it appears in a docker registry.
+  //
+  // When using GCR, this will likely be a fully-qualified URL and either an
+  // explicit tag or sha256 hash. For example, a Java worker image may be
+  // similar to one of the following:
+  //
+  //   - gcr.io/grpc-testing/java_worker:v3.2.19
+  //   - gcr.io/grpc-testing/java_worker@sha256:82b6360a84c19f23ed7ee9...
+  //
+  // The "latest" tag is automatically set by the registry, and there
+  // are no guarantees that it will point to a specific image. It should be
+  // avoided.
+  string container = 1;  // [(google.api.field_behavior) = REQUIRED];
+
+  // The kind which assigns the responsibilities of this container.
+  Kind kind = 2;  // [(google.api.field_behavior) = OPTIONAL];
+
+  // The number of identical components that should be created with these
+  // settings. For example, a "client" component can be configured once and
+  // specify "2" replicas. Two identical versions will be available for tests.
+  //
+  // This field is ignored for driver components. It may eventually be supported
+  // on server components, but this is not yet available.
+  int32 replicas = 3;
+}
+
+message StartTestSessionRequest {
+  // The test scenarios to run as a part of the test session.
+  repeated grpc.testing.Scenario scenarios = 1;  // [(google.api.field_behavior) = REQUIRED];
+
+  // The component which orchestrates running the test scenarios amongst workers.
+  Component driver = 2;  // [(google.api.field_behavior) = OPTIONAL];
+
+  // The components that are required as part of the test. Normally, this will
+  // involve at least one server and a number of clients.
+  repeated Component workers = 3;  // [(google.api.field_behavior) = REQUIRED];
+
+  // An optional location where the results should be written.
+  oneof result_destination {
+    // The name of the BigQuery table where results should be written.
+    string bq_results_table = 4;  // [(google.api.field_behavior) = OPTIONAL];
+
+    // The fully qualified GCS URI where a results file should be written.
+    string gcs_results_file_uri = 5;  // [(google.api.field_behavior) = OPTIONAL];
+  }
+ }
+
+message TestSessionResults {
+  // The results for each scenario.
+  repeated grpc.testing.ScenarioResult scenario_results = 1;
+
+  // The amount of time that this session lived, including all actions from
+  // scheduling to termination.
+  google.protobuf.Duration time_elapsed = 2;
+}
+
+message Event {
+  enum Kind {
+    // An infrastructure problem. This is outside of the test runner's control,
+    // and a bug should be filed if it is encountered.
+    INTERNAL = 0;
+
+    // Waiting for an executor to process the test session.
```

Comment:
the QUEUE, ACCEPT and PROVISION seem to be quite specific, not sure if it's actually useful to differentiate between those. You might wanna start will small number of events and add more if they prove to be useful.

Fine to leave as is for a prototype

Reply:
Noted. They are fine-grained. That said, the prototype has to monitor these events anyway, so there is little overhead. In the long term, they may provide some useful insights for idle jobs and improving your run times.  For example:

- Long QUEUE times lead us to ask, "do we not have enough resources in the cluster?"
- Long ACCEPT times tell us that our executors are not optimized (excessive time wasted before actually interacting with Kubernetes)
- Long PROVISION times tell us that Kubernetes or container images need to be optimized